nltk/tokenize/treebank.py

```diff
@@ -147,36 +147,46 @@ def tokenize(self, text, convert_parentheses=False, return_str=False):
-
             >>> from nltk.tokenize import TreebankWordTokenizer
-            >>> s = '''Good muffins cost $3.88\\nin New (York).  Please (buy) me\\ntwo of them.\\n(Thanks).'''
-            >>> expected = [(0, 4), (5, 12), (13, 17), (18, 19), (19, 23),
-            ... (24, 26), (27, 30), (31, 32), (32, 36), (36, 37), (37, 38),
-            ... (40, 46), (47, 48), (48, 51), (51, 52), (53, 55), (56, 59),
-            ... (60, 62), (63, 68), (69, 70), (70, 76), (76, 77), (77, 78)]
+            >>> s = '''I said, "I'd like to buy some ''good muffins" which cost $3.88\\n each in New (York)."'''
+            >>> expected = [(0, 1), (2, 6), (6, 7), (8, 9), (9, 10), (10, 12),
+            ... (13, 17), (18, 20), (21, 24), (25, 29), (30, 36), (37, 44),
+            ... (44, 45), (46, 51), (52, 56), (57, 58), (58, 62), (64, 68),
+            ... (69, 71), (72, 75), (76, 77), (77, 81), (81, 82), (82, 83),
+            ... (83, 84)]
             >>> TreebankWordTokenizer().span_tokenize(s) == expected
             True
-            >>> expected = ['Good', 'muffins', 'cost', '$', '3.88', 'in',
-            ... 'New', '(', 'York', ')', '.', 'Please', '(', 'buy', ')',
-            ... 'me', 'two', 'of', 'them.', '(', 'Thanks', ')', '.']
+            >>> expected = ['I', 'said', ',', '"', 'I', "'d", 'like', 'to',
+            ... 'buy', 'some', "''good", 'muffins', '"', 'which', 'cost', '$',
+            ... '3.88', 'each', 'in', 'New', '(', 'York', ')', '.', '"']
             >>> [s[start:end] for start, end in TreebankWordTokenizer().span_tokenize(s)] == expected
             True
 
         """
-        raw_tokens = self.tokenize(text)
-
         # Convert converted quotes back to original double quotes
-        # Do this only if original text contains double quote(s)
-        if '"' in text:
-            # Find double quotes and converted quotes
-            matched = [m.group() for m in re.finditer(r'[(``)(\'\')(")]+', text)]
-            
-            # Replace converted quotes back to double quotes
-            tokens = [matched.pop(0) if tok in ['"', "``", "''"] else tok for tok in raw_tokens]
-        else:
-            tokens = raw_tokens
-
-        return align_tokens(tokens, text)
+        ix = 0
+
+        spans = []
+        for word_token in self.tokenize(text):
+            if word_token in ('``', "''"):
+                orig_idx = text.find(word_token, ix)
+                quote_idx = text.find('"', ix)
+                if orig_idx < 0:
+                    real_token = '"'
+                elif quote_idx < 0:
+                    real_token = word_token
+                elif orig_idx < quote_idx:
+                    real_token = word_token
+                else:
+                    real_token = '"'
+            else:
+                real_token = word_token
+            ix = text.find(real_token, ix)
+            end = ix + len(real_token)
+            spans.append((ix, end))
+            ix = end
+
+        return spans
```

Comment:
Maybe it's better for this code to improve  `align_tokens()` at https://github.com/nltk/nltk/blob/develop/nltk/tokenize/util.py#L220 so that its more of a generic function and not just specific to the `TreebankTokenizer`

Reply:
Thanks. I think it is a good idea. However, I am not sure if the align_tokens() function will be used in other tokenizers' span_tokenize() function in the future? If other tokenizers have other unique behaviors that will modify "tokens" of the original string (similar to modifying " to ``. Is it possible to modify . . . (as one token) to ... in another tokenizer? This is just a made-up example). In that case align_tokens() will still break if such other tokenizer uses align_tokens().

I think the current align_tokens() should work well if a tokenizer does not modify any "tokens". I can find examples for such tokenizer, e.g. tokenize.util.string_span_tokenize(), tokenize.util.regexp_span_tokenize() - they are very simple tokenizers tho.

Adding my code to align_tokens() will not do harm in terms of quality, but it would add some computational cost for tokenizers which doesn't modify " to `` or ''.

Another possibility is to extend align_tokens() to take one additional optional parameter "modified_tokens", e.g. {'"': ['``', "''"]} to specify that some tokens may be modified to other forms.

I would like to hear your thought. Thank you!

Reply:
I was thinking more of replacing the whole `align_tokens()` with a better code like yours.

But I think there are there're better string search and offset indexing strategies that should be considered if possible. 

The ideal solution would have to be that the offsets are kept and updated during tokenization =(

Someone else with a better understanding of regex and strings should review this PR.

Reply:
Thank you @alvations ! I agree with you about the ideal solution where the offset should be kept during the tokenize() logic. 

Here are just my two cents:
I am not very familiar with the tokenizer regex code. But from what I can see, there are quite a few modifications (some of them are transient modifications that will be modified back to original state) in the tokenize() function of the  TreebankWordTokenizer. Therefore keeping the original offsets might require quite a lot of thought and new code - and it could even be computationally more expensive than my current method. But once it's done it should guarantee correctness.

If it sounds OK we can start with my current method as a workaround solution - I can also add some note in tokenize() method to update the span_tokenize() method just in case some other special tokens will be modified (like " => ``) in the future?